File: backend/internal/vision/detector.go
package vision

import (
	"context"
	"fmt"
	"image"

	"gocv.io/x/gocv"
)

type Detector struct {
	params Params
}

func NewDetector(params Params) *Detector {
	return &Detector{params: params}
}

func (d *Detector) Params() Params {
	return d.params
}

func (d *Detector) Detect(ctx context.Context, data []byte) ([]Box, error) {
	config, err := ValidateImage(data, d.params.MaxPixels)
	if err != nil {
		return nil, err
	}
	bounds := image.Rect(0, 0, config.Width, config.Height)
	// Decode the image to grayscale
	gray, err := gocv.IMDecode(data, gocv.IMReadGrayScale)
	if err != nil {
		return nil, fmt.Errorf("%w: %v", ErrCorrupt, err)
	}
	defer gray.Close()
	if gray.Empty() || gray.Cols() != config.Width || gray.Rows() != config.Height {
		return nil, ErrCorrupt
	}
	if err := ctx.Err(); err != nil {
		return nil, err
	}

	ink := gocv.NewMat()
	defer ink.Close()
	// Apply adaptive threshold to the grayscale image to create a binary image
	err = gocv.AdaptiveThreshold(
		gray,
		&ink,
		255,
		gocv.AdaptiveThresholdGaussian,
		gocv.ThresholdBinaryInv,
		d.params.AdaptiveBlockSize,
		d.params.AdaptiveC,
	)
	if err != nil {
		return nil, fmt.Errorf("adaptive threshold: %w", err)
	}
	if err := ctx.Err(); err != nil {
		return nil, err
	}
	// Create a ruling mask by keeping only pixels that belong to straight horizontal or vertical runs at least kernelLength long
	ruling, err := rulingMask(ink, d.params.LineKernelLength)
	if err != nil {
		return nil, err
	}
	defer ruling.Close()
	if err := ctx.Err(); err != nil {
		return nil, err
	}

	candidates := d.findCandidates(ink, ruling)
	return finalize(candidates, bounds, d.params.DedupeIoU), nil
}

// keeps only pixels that belong to straight horizontal or vertical lines of at least kernelLength length.
func rulingMask(ink gocv.Mat, kernelLength int) (gocv.Mat, error) {
	horizontalKernel := gocv.GetStructuringElement(gocv.MorphRect, image.Pt(kernelLength, 1))
	defer horizontalKernel.Close()
	verticalKernel := gocv.GetStructuringElement(gocv.MorphRect, image.Pt(1, kernelLength))
	defer verticalKernel.Close()
	closeKernel := gocv.GetStructuringElement(gocv.MorphRect, image.Pt(3, 3))
	defer closeKernel.Close()

	horizontal := gocv.NewMat()
	defer horizontal.Close()
	vertical := gocv.NewMat()
	defer vertical.Close()
	combined := gocv.NewMat()
	defer combined.Close()

	if err := gocv.MorphologyEx(ink, &horizontal, gocv.MorphOpen, horizontalKernel); err != nil {
		return gocv.Mat{}, fmt.Errorf("horizontal opening: %w", err)
	}
	if err := gocv.MorphologyEx(ink, &vertical, gocv.MorphOpen, verticalKernel); err != nil {
		return gocv.Mat{}, fmt.Errorf("vertical opening: %w", err)
	}
	if err := gocv.BitwiseOr(horizontal, vertical, &combined); err != nil {
		return gocv.Mat{}, fmt.Errorf("combine rulings: %w", err)
	}

	ruling := gocv.NewMat()
	if err := gocv.MorphologyEx(combined, &ruling, gocv.MorphClose, closeKernel); err != nil {
		ruling.Close()
		return gocv.Mat{}, fmt.Errorf("close rulings: %w", err)
	}
	return ruling, nil
}

File: backend/src/homevision/vision/detector.py
import cv2
import numpy as np
from cv2.typing import MatLike

from homevision.vision.boxes import Box, Rect, finalize
from homevision.vision.candidates import find_candidates
from homevision.vision.image import CorruptImageError, validate_image
from homevision.vision.params import Params


class Detector:
    """Finds checkboxes in a scanned form page and classifies each as checked or not."""

    def __init__(self, params: Params | None = None) -> None:
        self.params = params or Params()

    def detect(self, data: bytes) -> list[Box]:
        """Return the checkboxes in a PNG or JPEG, in reading order.

        Raises ImageError for unsupported, oversized, or undecodable input.
        """
        validate_image(data, self.params.max_pixels)

        # Decode the image to grayscale
        gray = cv2.imdecode(np.frombuffer(data, dtype=np.uint8), cv2.IMREAD_GRAYSCALE)
        if gray is None or gray.size == 0:
            raise CorruptImageError
        height, width = gray.shape

        # Apply adaptive threshold to the grayscale image to create a binary image
        ink = cv2.adaptiveThreshold(
            gray,
            255,
            cv2.ADAPTIVE_THRESH_GAUSSIAN_C,
            cv2.THRESH_BINARY_INV,
            self.params.adaptive_block_size,
            self.params.adaptive_c,
        )
        ruling = ruling_mask(ink, self.params.line_kernel_length)

        candidates = find_candidates(ink, ruling, self.params)
        return finalize(candidates, Rect(0, 0, width, height), self.params.dedupe_iou)


def ruling_mask(ink: MatLike, kernel_length: int) -> MatLike:
    """Keep only pixels on straight horizontal or vertical runs of at least kernel_length."""
    horizontal_kernel = cv2.getStructuringElement(cv2.MORPH_RECT, (kernel_length, 1))
    vertical_kernel = cv2.getStructuringElement(cv2.MORPH_RECT, (1, kernel_length))
    close_kernel = cv2.getStructuringElement(cv2.MORPH_RECT, (3, 3))

    horizontal = cv2.morphologyEx(ink, cv2.MORPH_OPEN, horizontal_kernel)
    vertical = cv2.morphologyEx(ink, cv2.MORPH_OPEN, vertical_kernel)
    combined = cv2.bitwise_or(horizontal, vertical)
    return cv2.morphologyEx(combined, cv2.MORPH_CLOSE, close_kernel)
